MapleServer2/Packets/ShopPacket.cs

```diff
@@ -0,0 +1,161 @@
+            pWriter.WriteInt(shop.Id);
+            pWriter.WriteLong(shop.NextRestock); // timestamp for next restock
+            pWriter.WriteInt();
+            pWriter.WriteShort(15);
+            pWriter.WriteInt(shop.Category); // ShopCategory (916 = Juice)
+            pWriter.WriteBool(false); //
+            pWriter.WriteBool(shop.RestrictSales); // restrict sales (default: false)
+            pWriter.WriteBool(false); // shop can be restocked
+            pWriter.WriteBool(false);
+            pWriter.WriteBool(false);
+            pWriter.WriteBool(shop.AllowBuyback); // show buyback tab (default: true)
+            pWriter.WriteBool(false);
+            pWriter.WriteBool(true); // unknown
+            pWriter.WriteBool(false);
+            pWriter.WriteMapleString(shop.Name); // shopName
+            
+            return pWriter;
+        }
+
+        public static Packet Reload()
+        {
+            PacketWriter pWriter = PacketWriter.Of(SendOp.SHOP);
+            pWriter.WriteByte((byte) ShopMode.Reload);
+            pWriter.WriteByte();
+            pWriter.WriteByte();
+            
+            return pWriter;
+        }
+
+        public static Packet Close()
+        {
+            PacketWriter pWriter = PacketWriter.Of(SendOp.SHOP);
+            pWriter.WriteShort();
+            
+            return pWriter;
+        }
+
+        public static Packet Buy(int itemId, int quantity, int price, byte currencyType)
+        {
+            PacketWriter pWriter = PacketWriter.Of(SendOp.SHOP);
+            pWriter.WriteByte((byte) ShopMode.Buy);
+            pWriter.WriteInt(itemId); // Item ID
+            pWriter.WriteInt(quantity); // Quantity
+            pWriter.WriteInt(price * quantity); // Total price
+            pWriter.WriteShort(currencyType); // Currency type
+            
+            return pWriter;
+        }
+
+        public static Packet Sell(long itemUid, int itemId, int quantity)
+        {
+            PacketWriter pWriter = PacketWriter.Of(SendOp.SHOP);
+            pWriter.WriteByte((byte) ShopMode.Sell);
+            pWriter.WriteInt(quantity);
+            pWriter.WriteShort();
+            pWriter.WriteInt(itemId);
+            pWriter.WriteByte(1);
+            pWriter.WriteByte(1);
+            pWriter.WriteByte();
+            pWriter.WriteInt();
+            
+            // Write item data
+            pWriter.WriteByte();
+            pWriter.WriteByte();
+            pWriter.WriteInt(1); // Unknown (amount?)
+            pWriter.WriteInt(-1);
+            pWriter.WriteLong(DateTimeOffset.UtcNow.ToUnixTimeSeconds()); // Item creation time
+            pWriter.WriteZero(56);
+            pWriter.WriteInt(-1);
+            pWriter.WriteZero(69);
+            pWriter.WriteInt(1);
+            pWriter.WriteZero(14);
+            pWriter.WriteInt(6);
+            pWriter.WriteZero(10);
+            pWriter.WriteInt(1);
+            pWriter.WriteZero(18);
```

Comment:
Can't this be `pWriter.WriteItem(item)`? It looks like the same logic.

Reply:
I thought so too, but it's slightly different for shops unfortunately... (thanks Nexon...)